src/sentry/taskworker/scheduler.py

```diff
@@ -139,12 +141,130 @@ def runtime_after(self, start: datetime) -> datetime:
+        return (
+            f"{self.minute} {self.hour} {self.day_of_month} {self.month_of_year} {self.day_of_week}"
+        )
+
+
+class CrontabSchedule(Schedule):
+    """
+    Task schedules defined as crontab expressions.
+
+    Last run state is used to determine the next scheduled time.
+
+    If last run is in the future, the future value will be used to
+    calculate the next scheduled time.
+
+    If runs are missed, the schedule will align to the next interval
+    that would be scheduled. Lost runs are not recovered.
+    """
+
+    def __init__(self, name: str, crontab: crontab) -> None:
+        self._crontab = crontab
+        self._name = name
+        try:
+            self._cronsim = CronSim(str(crontab), timezone.now())
+        except CronSimError as e:
+            raise ValueError(f"crontab expression {self._crontab} is invalid") from e
+
+    def is_due(self, last_run: datetime | None = None) -> bool:
+        """Check if the schedule is due to run again based on last_run."""
+        if last_run is None:
+            return True
+        remaining = self.remaining_seconds(last_run)
+        return remaining <= 0
+
+    def remaining_seconds(self, last_run: datetime | None = None) -> int:
+        """
+        Get the number of seconds until this schedule is due again
+
+        Use the current time to find the next schedule time
+        """
+        if last_run is None:
+            return 0
+
+        # This could result in missed beats, or increased load on redis.
+        last_run = last_run.replace(second=0, microsecond=0)
+        now = timezone.now().replace(second=0, microsecond=0)
+
+        # A future last_run means we should wait until the
+        # next scheduled time, and then we can try again.
+        # we could be competing with another scheduler, or
+        # missing beats.
+        if last_run > now:
+            logger.warning(
+                "taskworker.scheduler.future_value",
+                extra={
+                    "task": self._name,
+                    "last_run": last_run,
+                    "now": now,
+                },
+            )
+            next_run = self._advance(last_run + timedelta(minutes=1))
+            return int(next_run.timestamp() - now.timestamp())
+
+        # If last run is in the past, see if the next runtime
+        # is in the future.
+        if last_run < now:
+            next_run = self._advance(last_run)
+            # Our next runtime is in the future, or now
+            if next_run >= now:
+                return int(next_run.timestamp() - now.timestamp())
+
+            # still in the past, we missed an interval :(
+            next_run = self._advance(now)
+            logger.warning(
+                "taskworker.scheduler.missed_interval",
+                extra={
+                    "task": self._name,
+                    "last_run": last_run,
+                    "now": now,
+                    "next_run": next_run,
+                },
```

Comment:
Would it be possible even after another `self._advance`, we are still in the past? E.g. we missed 2 or more intervals. In this case, would we want to keep advancing until next_run is at a future time?

Reply:
> Would it be possible even after another self._advance, we are still in the past? 

Yes, we could miss multiple intervals. I didn't want to add a loop here, so if the first advance is still in the past, the second `_advance()` call on 229 uses `now` so that the second `_advance` will happen in the future.